server/protocols/jmap/src/main/java/org/apache/james/jmap/event/PropagateLookupRightListener.java

```diff
@@ -0,0 +1,130 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.event;
+
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.acl.ACLDiff;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxACL.Entry;
+import org.apache.james.mailbox.model.MailboxACL.Right;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.event.EventFactory.MailboxRenamedEventImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PropagateLookupRightListener implements MailboxListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropagateLookupRightListener.class);
+
+    private final StoreRightManager storeRightManager;
+    private final MailboxSessionMapperFactory mapperFactory;
+
+    @Inject
+    public PropagateLookupRightListener(StoreRightManager storeRightManager, MailboxSessionMapperFactory mapperFactory) {
+        this.storeRightManager = storeRightManager;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public ListenerType getType() {
+        return ListenerType.ONCE;
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return ExecutionMode.ASYNCHRONOUS;
+    }
+
+    @Override
+    public void event(Event event) {
+        MailboxSession mailboxSession = event.getSession();
+
+        if (event instanceof MailboxACLUpdated) {
+            MailboxACLUpdated aclUpdateEvent = (MailboxACLUpdated) event;
+
+            updateLookupRightOnParent(mailboxSession, event.getMailboxPath(), aclUpdateEvent.getAclDiff());
+        } else if (event instanceof MailboxRenamed) {
+            MailboxRenamedEventImpl renamedEvent = (MailboxRenamedEventImpl) event;
+
+            updateLookupRightOnParent(mailboxSession, renamedEvent.getNewPath(), renamedEvent.getMailbox().getACL());
+        }
+    }
+
+    private void updateLookupRightOnParent(MailboxSession mailboxSession, MailboxPath mailboxPath, MailboxACL mailboxACL) {
+        getParent(mailboxPath, mailboxSession)
```

Comment:
`getParent().forEach()` doesn't make sense to me. do the mailbox has several parents ?

Reply:
Maybe getAllHierarchyLevels? Like A.B.C then we need to update A and A.B

Reply:
I changed to `listAncestors`